### PR TITLE
[14.0][IMP] auth_oidc: Logout the user from the auth provider

### PR DESCRIPTION
**Pull request by wluyima**
Added a mechanism to logout the user from the auth provider when

### CODE REVIEW

File: auth_oidc/controllers/main.py
Comment (ap-wtioit) [inline auth_oidc/controllers/main.py]: the logout_endpoint may already contain query parameters:
> 
    end_session_endpoint
        REQUIRED. URL at the OP to which an RP can perform a redirect to request that the End-User be logged out at the OP. This URL MUST use the https scheme and MAY contain port, path, and query parameter components. 

https://openid.net/specs/openid-connect-rpinitiated-1_0.html

So it would be better to not add a hardcocded questionmark here but use a more dynamic solution to adding parameters.

Also it seems suspicious that the redirect url `r` is not encoded here. (But i have not tested it yet)

File: auth_oidc/models/auth_oauth_provider.py
Comment (ap-wtioit) [inline auth_oidc/models/auth_oauth_provider.py]: I guess this is expected to be the same value as `end_session_endpoint` from the provider discovery metadata, can you add a remark to the documentation that this is the value to be used.

**Review by ap-wtioit — CHANGES_REQUESTED** [review]
Thanks for your efforts to make this module better. Could you run the OCA pre-commit or incoperate the suggestions from the log output: https://github.com/OCA/server-auth/actions/runs/9479497854/job/26118147534?pr=666

File: auth_oidc/controllers/main.py
Comment (ap-wtioit) [reply auth_oidc/controllers/main.py]: Yes `r` should be encoded when added as post_logout_redirect_uri. (e.g. when redirect is `/web/login?db=devel&something=different` it would possibly end up as parameter something on the logout_endpoint)

Comment (wluyima) [reply auth_oidc/controllers/main.py]: Makes sense, I will update the PR

File: auth_oidc/models/auth_oauth_provider.py
Comment (wluyima) [reply auth_oidc/models/auth_oauth_provider.py]: Sounds good, I could also rename the property to `end_session_endpoint`

File: auth_oidc/controllers/main.py
Comment (wluyima) [inline auth_oidc/controllers/main.py]: @ap-wtioit is redirect expected to always be relative to the base URL? If full URLs are supported then the line above would need to be updated to be optional if redirect starts with `http` or `https`

Comment (ap-wtioit) [reply auth_oidc/controllers/main.py]: i think only relative redirects are used in odoo but there is no technical limit for that and the redirect paramter is sent directly to werkzeug: https://github.com/odoo/odoo/blob/14.0/addons/web/controllers/main.py#L1319

so yes that's a good catch and web.base.url should only be used when the url is not starting with a protocol (i guess http(s) will suffice), also there is the possibility that the redirect is relative to /web/session/logout (e.g. `../login`, or `destroy`) but i think this is also not used currently (and i tested it only locally for checking security)

Comment (wluyima) [reply auth_oidc/controllers/main.py]: You mentioned `destroy`, I see it also logs out the user, do we need to override it too to logout the user in the IDP?

Comment (ap-wtioit) [reply auth_oidc/controllers/main.py]: i can only see `destroy` being used in one Odoo Enterprise application "project_timesheet_synchro / Synchronization with the external timesheet application" and that code upon login does not support the things added by auth_oidc so i would skip this usecase here.

Comment (wluyima) [reply auth_oidc/controllers/main.py]: So is this good to go?

Comment (ap-wtioit) [reply auth_oidc/controllers/main.py]: I'd like to test it again (most likely tomorrow) but yes this looks good to me now.

File: auth_oidc/controllers/main.py
Comment (wluyima) [inline auth_oidc/controllers/main.py]: I have sneaked in support for full URLs

Comment (ap-wtioit) [reply auth_oidc/controllers/main.py]: Please use `not r.startswith("http:") and not r.startswith("https:")` here. the check you used would skip adding the base url for all redirects containing http.

**Comment by ap-wtioit** [comment]
Reviewed again, did you see my comment regarding oca-precommit? The check is still failing.

**Comment by wluyima** [comment]
@ap-wtioit when do you plan to have this merged?

File: auth_oidc/controllers/main.py
Comment (wluyima) [reply auth_oidc/controllers/main.py]: Done

**Review by ap-wtioit — APPROVED** [review]

**Comment by ap-wtioit** [comment]
> @ap-wtioit when do you plan to have this merged?

to me it looks good now. to get it merged we need a second approval, ideally from a PSC (Project Steering Committee member) or the maintainer of the module. then it can be merged (by the maintainer or PSC).

**Comment by wluyima** [comment]
@sbidoul do you mind becoming the second reviewer/approver on this PR?

**Comment by sbidoul** [comment]
I will look into it. Not had time yet.

**Comment by wluyima** [comment]
I see that this was added to the 14.0 milestone, it would be great to forward port the changes to later milestones too i.e. 15, 16 and 17

**Review by wtaferner — APPROVED** [review]

File: auth_oidc/models/auth_oauth_provider.py
Comment (sbidoul) [inline auth_oidc/models/auth_oauth_provider.py]: I'd suggest naming this field `end_session_endpoint` to match the field in the openid configuration document. This will make the code easier to understand.

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]: Let's use `urllib.parse.urljoin` instead of `+` here. It will be better in case `web.base.url` ends with a `/`.

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]: I'm not too comfortable with parsing a URL this way. What if, for instance, there is more than 1 `?` in the url. It *should* not happen of course, but... It would be more robust to use `urllib.parse.urlparse` and `parse_qs`.

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]: It might be nice to extract this url manipulation logic in a little class method that could be unit tested.

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]: A little detail: let's use explicit variable names?

File: auth_oidc/models/auth_oauth_provider.py
Comment (sbidoul) [inline auth_oidc/models/auth_oauth_provider.py]: Shall we mention that when this field is set, logging out of Odoo also causes a complete logout of the identify provider?

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py:60]: Can you add a comment or docstring referencing the spec that this implements? 
https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout

**Review by sbidoul — CHANGES_REQUESTED** [review]
Hi,

Thanks for this contrib.

I made a few comments, the most important being the URL parsing logic that should use urllib.parse for robustness. I think this is important is such a security sensitive module.

Could you also add a test? The existing tests should give inspiration.

Head up about forward porting: I tested this on 16.0 and it will require some work to adapt, because Odoo now forces a local redirect (requests.redirect has a local=True argument):

https://github.com/odoo/odoo/blob/e889bd2c993a456d3077beaa3271f73e3510665e/addons/web/controllers/session.py#L85-L87

File: auth_oidc/controllers/main.py
Comment (wluyima) [reply auth_oidc/controllers/main.py]: Is there documentation on how to run the test suite?

Comment (sbidoul) [reply auth_oidc/controllers/main.py]: @wluyima This [part of the odoo documentation](https://www.odoo.com/documentation/17.0/developer/reference/backend/testing.html?highlight=test%20suite#running-tests) should get you started. Let me know if you hit any issue.

Comment (wluyima) [reply auth_oidc/controllers/main.py]: Thanks @sbidoul is there a way I can find documentation to run Odoo 14.0 locally from source? I was following the [15.0 documentation](https://www.odoo.com/documentation/15.0/administration/on_premise/source.html) and I can't seem to install the packages in the requirements.txt file, I get [this error](https://dpaste.com/8A7SVGMGA) when I try to install the packages, do you have any clues? I'm using Python **3.12.4**. I'm not sure if is not compatible, is there documentation on Odoo and Python versions compatibility?

Comment (sbidoul) [reply auth_oidc/controllers/main.py]: For Odoo 14 I generally use python 3.8. It will for sure not work with 3.12.

I recommend installing the Odoo dependencies in a virtual environment:

```
python3.8 -m venv ./venv
source ./venv/bin/activate
pip install -r https://raw.githubusercontent.com/odoo/odoo/14.0/requirements.txt
```

But if you don't use Odoo 14, why do you do this PR to the 14.0 branch?

Comment (wluyima) [reply auth_oidc/controllers/main.py]: Thanks @sbidoul! I did say we're using Odoo 14 which is why this PR is issued against the 14.0 branch and I'm installing the dependencies in a virtual environment but I was using python 3.12, I'll try with python 3.8 and see if it works. I was actually able to later find the documentation for Odoo 14 [here](https://www.odoo.com/documentation/14.0/administration/install/source.html).

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]: This test is now unnecessary, I think, because urljoin will do the right thing.

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]:
```suggestion
                post_logout_redirect_uri = redirect
```

**Comment by wluyima** [comment]
@sbidoul tests have been added

**Comment by wluyima** [comment]
@sbidoul @ap-wtioit tests were added, did you get a chance to take a look?

**Review by gbrito — APPROVED** [review]

File: auth_oidc/controllers/main.py
Comment (sbidoul) [inline auth_oidc/controllers/main.py]: When I see `red_url` my brain wonders why that url is not blue ;)
Let's not abuse abbreviations.

File: auth_oidc/models/auth_oauth_provider.py
Comment (sbidoul) [reply auth_oidc/models/auth_oauth_provider.py]: Can you also finish that sentence with a `.` as for other help texts.

File: auth_oidc/tests/test_auth_oidc_logout.py
Comment (sbidoul) [inline auth_oidc/tests/test_auth_oidc_logout.py]: This assert is useless as None is always falsy, and if you remove it, this test looks identical to the previous one?

**Review by sbidoul — COMMENTED** [review]
A couple of minor remarks, otherwise looks good now.

Can you squash commits?

File: auth_oidc/tests/test_auth_oidc_logout.py
Comment (wluyima) [reply auth_oidc/tests/test_auth_oidc_logout.py]: Good catch! It was an oversight.

The tests are not identical, the previous test is for the scenario where a user is not associated to a provider, while this test is for the scenario where no end_session_endpoint is configured for the provider, the difference is on lines 58 and 66.

**Comment by wluyima** [comment]
Done and squashed commits

**Review by sbidoul — APPROVED** [review]
Thanks!

**Comment by wluyima** [comment]
Hi @ap-wtioit, is this waiting on anything to be merged?

**Comment by ap-wtioit** [comment]
@wluyima a maintainer needs to merge it, preferably @sbidoul. but since he approved it already maybe he just forgot.

**Comment by sbidoul** [comment]
Indeed.

/ocabot merge minor

Thanks for the work @wluyima

**Pull request merged**
